shared/src/permissions.rs

```diff
@@ -0,0 +1,157 @@
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+use strum::IntoEnumIterator;
+
+#[derive(Clone, Copy, Hash, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Permission {
+    Account(AccountPermission),
+    Universe(GenericPermission),
+}
+
+#[derive(
+    Clone,
+    Copy,
+    Hash,
+    Eq,
+    PartialEq,
+    Debug,
+    Serialize,
+    Deserialize,
+    strum_macros::EnumString,
+    strum_macros::ToString,
+    strum_macros::EnumIter,
+)]
+pub enum AccountPermission {
+    PermanentBan,
+    TemporaryBan,
+    List,
+    View,
+}
+
+impl ToString for Permission {
+    fn to_string(&self) -> String {
+        let (service, permission) = match self {
+            Permission::Account(perm) => (Service::Account, perm.to_string()),
+            Permission::Universe(perm) => (Service::Universe, perm.to_string()),
+        };
+
+        format!("{}({})", service.to_string(), permission)
+    }
+}
+
+impl FromStr for Permission {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut service = String::new();
+        let mut permission = String::new();
+        enum State {
+            InService,
+            InPermission,
+            AtEnd,
+        }
+        let mut state = State::InService;
+
+        for c in s.chars() {
+            match c {
+                '(' => {
+                    state = State::InPermission;
+                }
+                ')' => {
+                    state = State::AtEnd;
+                }
+                _ => match state {
+                    State::InService => service.push(c),
+                    State::InPermission => permission.push(c),
+                    State::AtEnd => anyhow::bail!("junk found after close paren"),
+                },
+            }
+        }
+
+        let service = Service::from_str(&service)?;
+
+        let permission = match service {
+            Service::Account => Permission::Account(AccountPermission::from_str(&permission)?),
+            Service::Universe => Permission::Universe(GenericPermission::from_str(&permission)?),
+        };
+
+        Ok(permission)
+    }
+}
+
+#[derive(
+    Clone,
+    Copy,
+    Hash,
+    Eq,
+    PartialEq,
+    Debug,
+    Serialize,
+    Deserialize,
+    strum_macros::EnumString,
+    strum_macros::ToString,
+    strum_macros::EnumIter,
+)]
+pub enum GenericPermission {
+    List,
+    View,
+}
+
+#[derive(
+    Hash,
+    Eq,
+    PartialEq,
+    Debug,
+    Serialize,
+    Deserialize,
+    strum_macros::EnumString,
+    strum_macros::ToString,
+    strum_macros::EnumIter,
+)]
+pub enum Service {
+    Account,
+    Universe,
+}
+
+impl Service {
+    pub fn permissions(&self) -> Vec<Permission> {
+        match self {
+            Service::Account => AccountPermission::iter().map(Permission::Account).collect(),
+            Service::Universe => GenericPermission::iter()
+                .map(Permission::Universe)
+                .collect(),
```

Comment:
These could be `Lazy`, wish `strum` had something for that.

Reply:
I decided the static cache in the new implementation wasn't worth it. These loops are limited to just the values in question. In the previous implementation, the loop would always need to go over all values, which is why I had Lazy in here originally. If I were to redo this from the current implementation, I'd expose a custom iterator, not reintroduce the lazy. Just seemed like overkill for the frequency this is being called (only upon signin once)